src/Entities/Entity.cpp

```diff
@@ -1559,68 +1536,81 @@ bool cEntity::DetectPortal()
+	#ifdef _DEBUG
+		// Take note of old chunk coords
+		auto OldChunkCoords = cChunkDef::BlockToChunk(GetPosition());
+	#endif
+
+	// Set position to the new position
+	SetPosition(a_WorldChangeInfo.m_NewPosition);
+
+	if (a_WorldChangeInfo.m_SetPortalCooldown)
 	{
-		// A Plugin doesn't allow the entity to changing the world
-		return false;
+		m_PortalCooldownData.m_TicksDelayed = 0;
+		m_PortalCooldownData.m_ShouldPreventTeleportation = true;
+	}
+
+	if (GetWorld() == a_WorldChangeInfo.m_NewWorld)
+	{
+		// Moving to same world, don't need to remove from world
+		return;
 	}
 
 	// Stop ticking, in preperation for detaching from this world.
 	SetIsTicking(false);
 
-	// Tell others we are gone
-	GetWorld()->BroadcastDestroyEntity(*this);
+	a_WorldChangeInfo.m_NewWorld->BroadcastSpawnEntity(*this);
 
-	// Take note of old chunk coords
-	auto OldChunkCoords = cChunkDef::BlockToChunk(GetPosition());
+	// Add to new world and removal from the old one
+	LOGD("Warping entity #%i (%s) from world \"%s\" to \"%s\". Source chunk: (%d, %d) ",
+		this->GetUniqueID(), this->GetClass(),
+		m_World->GetName().c_str(), a_WorldChangeInfo.m_NewWorld->GetName().c_str(),
+		OldChunkCoords.m_ChunkX, OldChunkCoords.m_ChunkZ
+	);
+	a_WorldChangeInfo.m_NewWorld->AddEntity(m_World->RemoveEntity(*this));
+	cRoot::Get()->GetPluginManager()->CallHookEntityChangedWorld(*this, *m_World);
+}
 
-	// Set position to the new position
-	SetPosition(a_NewPosition);
 
-	// Stop all mobs from targeting this entity
-	// Stop this entity from targeting other mobs
-	if (this->IsMob())
-	{
-		cMonster * Monster = static_cast<cMonster*>(this);
-		Monster->SetTarget(nullptr);
-		Monster->StopEveryoneFromTargetingMe();
-	}
 
-	// Queue add to new world and removal from the old one
-	cWorld * OldWorld = GetWorld();
-	SetWorld(a_World);  // Chunks may be streamed before cWorld::AddPlayer() sets the world to the new value
-	OldWorld->QueueTask([this, OldChunkCoords, a_World](cWorld & a_OldWorld)
-	{
-		LOGD("Warping entity #%i (%s) from world \"%s\" to \"%s\". Source chunk: (%d, %d) ",
-			this->GetUniqueID(), this->GetClass(),
-			a_OldWorld.GetName().c_str(), a_World->GetName().c_str(),
-			OldChunkCoords.m_ChunkX, OldChunkCoords.m_ChunkZ
-		);
-		a_World->AddEntity(a_OldWorld.RemoveEntity(*this));
-		cRoot::Get()->GetPluginManager()->CallHookEntityChangedWorld(*this, a_OldWorld);
-	});
-	return true;
-}
 
 
+bool cEntity::MoveToWorld(cWorld * a_World, Vector3d a_NewPosition, bool a_SetPortalCooldown, bool a_ShouldSendRespawn)
+{
+	ASSERT(a_World != nullptr);
 
+	// Ask the plugins if the entity is allowed to change world
+	if (cRoot::Get()->GetPluginManager()->CallHookEntityChangingWorld(*this, *a_World))
+	{
+		// A Plugin isn't allowing the entity to change world
+		return false;
+	}
 
+	if (!IsWorldChangeScheduled())
+	{
+		// Schedule a new world change.
+		m_WorldChangeInfo = cpp14::make_unique<sWorldChangeInfo>();
```

Comment:
This is thread-unsafe. One thread may set the `m_WorldChangeInfo` to a valid value while another thread is between the `if` above and this line.

Reply:
Is the new atomic version okay?

Reply:
I think conceivably https://github.com/cuberite/cuberite/pull/4004/files/#diff-f312969b3897e838a1085ffb60b7c4dfR1612 could be called between https://github.com/cuberite/cuberite/pull/4004/files/#diff-f312969b3897e838a1085ffb60b7c4dfR1604 and the if statement, which would cause a WorldChangeInfo to be scheduled twice in a row. I don't think that'd actually cause any problems, though, since most of the failsafes are in DoMoveToWorld.

Reply:
Both exchanges are atomic and the comparison is done on the exchange result and not from an additional load of `m_WorldChangeInfo` so I don't think that's an issue.